eslint.config.mjs

```diff
@@ -0,0 +1,416 @@
+        'error',
+        {
+          ignore: [
+            'ts-check',
+            'ts-expect-error',
+            'ts-ignore',
+            'ts-nocheck',
+            'typescript-to-proptypes-ignore',
+          ],
+        },
+      ],
+      // Fixes false positive when using both `inputProps` and `InputProps` on the same example
+      // See https://stackoverflow.com/questions/42367236/why-am-i-getting-this-warning-no-duplicate-props-allowed-react-jsx-no-duplicate
+      // TODO move to @mui/monorepo/.eslintrc
+      // TODO Fix <Input> props names to not conflict
+      'react/jsx-no-duplicate-props': [1, { ignoreCase: false }],
+      // TODO move to @mui/monorepo/.eslintrc, these are false positive
+      'react/no-unstable-nested-components': ['error', { allowAsProps: true }],
+    },
+  },
+  // Test start
+  {
+    files: [
+      // matching the pattern of the test runner
+      `**/*.test.${EXTENSION_TS}`,
+    ],
+    extends: createTestConfig({ useMocha: false }),
+    ignores: ['test/e2e/**/*', 'test/regressions/**/*'],
+    rules: {
+      'testing-library/no-container': 'off',
+    },
+  },
+  baseSpecRules,
+
+  {
+    files: [`**/*.test.${EXTENSION_TS}`, 'test/**'],
+    rules: {
+      'no-restricted-imports': [
+        'error',
+        {
+          paths: ['@testing-library/react', 'test/utils/index'],
+        },
+      ],
+    },
+  },
+
+  {
+    files: [
+      'packages/x-data-grid/**/*{.tsx,.ts,.js}',
+      'packages/x-data-grid-pro/**/*{.tsx,.ts,.js}',
+      'packages/x-data-grid-premium/**/*{.tsx,.ts,.js}',
+      'docs/src/pages/**/*.tsx',
+    ],
+    rules: {
+      'mui-x/no-direct-state-access': 'error',
+    },
+    languageOptions: {
+      parserOptions: {
+        tsconfigRootDir: dirname,
+        project: ['./tsconfig.json'],
+      },
+    },
+  },
+
+  // TODO remove, shouldn't disable prop-type generation rule.
+  // lot of public components are missing it.
+  {
+    files: ['**/*.tsx'],
+    ignores: ['**/*.spec.tsx'],
+    rules: {
+      'react/prop-types': 'off',
+    },
+  },
+
+  {
+    files: [`packages/*/src/**/*.${EXTENSION_TS}`],
+    ignores: ['**/*.d.ts', `**/*.spec.${EXTENSION_TS}`, `**/*.test.${EXTENSION_TS}`],
+    rules: {
+      'material-ui/mui-name-matches-component-name': [
+        'error',
+        {
+          customHooks: [
+            'useDatePickerProcessedProps',
+            'useDatePickerDefaultizedProps',
+            'useTimePickerDefaultizedProps',
+            'useDateTimePickerDefaultizedProps',
+            'useDateRangePickerDefaultizedProps',
+            'useDateTimeRangePickerDefaultizedProps',
+            'useTimeRangePickerDefaultizedProps',
+            'useDateCalendarDefaultizedProps',
+            'useMonthCalendarDefaultizedProps',
+            'useYearCalendarDefaultizedProps',
+            'useDateRangeCalendarDefaultizedProps',
+          ],
+        },
+      ],
+      'material-ui/disallow-react-api-in-server-components': 'error',
```

Comment:
This refactoring and investigation opened a "can of worms". 🙈 
https://github.com/mui/mui-x/pull/18570